pkg/controller/backup_deletion_controller.go

```diff
@@ -324,7 +327,14 @@ func (c *backupDeletionController) processRequest(req *velerov1api.DeleteBackupR
 
 	if features.IsEnabled(velerov1api.CSIFeatureFlag) {
 		log.Info("Removing CSI volumesnapshots")
-		if csiErrs := deleteCSIVolumeSnapshots(backup, c.csiSnapshotLister, c.csiSnapshotClient.SnapshotV1beta1(), log); len(csiErrs) > 0 {
+		if csiErrs := deleteCSIVolumeSnapshots(backup.Name, c.csiSnapshotLister, c.csiSnapshotClient.SnapshotV1beta1(), log); len(csiErrs) > 0 {
+			for _, err := range csiErrs {
+				errs = append(errs, err.Error())
+			}
+		}
+
+		log.Info("Removing CSI volumesnapshotcontents")
+		if csiErrs := deleteCSIVolumeSnapshotContents(backup.Name, c.csiSnapshotContentLister, c.csiSnapshotClient.SnapshotV1beta1(), log); len(csiErrs) > 0 {
```

Comment:
I wonder if we're going to run into any race conditions between the deletion of the VolumeSnapshots, which will trigger the CSI snapshot controller to go delete the related VolumeSnapshotContents, and our explicit deletion of the VolumeSnapshotContents themselves. Off the top of my head it seems like everything would probably work OK, but hard to say without testing. Did you observe any such issues?

Reply:
This is the test that I ran.
1. Deploy 4 instances of the [hostpath-plugin example](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/examples) each in their own namesapces.  
2. `v backup create b1 --include-namespaces csi-app,csi-app-1,csi-app-2,csi-app-3 --wait`
3. `v delete backup b1`
This is what I saw:
- `deleteCSIVolumeSnapshots` function:
    - set the `DeletionPolicy` on all the VSCs
    - deleted the VS 
- By the time `deleteCSIVolumeSnapshotContents` function was invoked the VSCs weren't all deleted yet. So:
- csiVSCLister returned 4 VSCs to be deleted. 
- the `setVolumeSnapshotContentDeletionPolicy` was called for each of the 4 VSCs. However, it returned with a `NotFound` error for the last 3 of the VSCs which we handle [here](https://github.com/vmware-tanzu/velero/pull/2480/files#diff-afe709fa1785e08707ca0825e3f60e68R527) to `continue` the deletion loop.
- Only for the VSC that `setVolumeSnapshotContentDeletionPolicy` didn't return the `NotFound` error we call the delete function on the VSC.

So I didn't see any errors. However, I think we should be handling the `NotFound` error returned from
`` `
err = csiClient.VolumeSnapshotContents().Delete(snapCont.Name, &metav1.DeleteOptions{})
``` 
I am going to make that change

Reply:
@ashish-amarnath 404 would essentially be a no-op, right?

Reply:
Yeah. We might run into that when we attempt to delete something that doesn't exist.
I am not sure if the `Delete(...)` handles that and returns nil on `NotFound`